torch/csrc/distributed/c10d/control_plane/WaitCounterHandler.cpp

```diff
@@ -0,0 +1,138 @@
+#include <torch/csrc/distributed/c10d/control_plane/WaitCounterHandler.hpp>
+
+#include <atomic>
+#include <chrono>
+#include <memory>
+#include <string>
+#include <unordered_map>
+
+#include <c10/util/Synchronized.h>
+#include <c10/util/WaitCounter.h>
+#include <nlohmann/json.hpp>
+
+namespace c10d {
+namespace control_plane {
+
+namespace {
+
+// Data structure to hold counter metrics
+struct CounterData {
+  std::atomic<int64_t> active_count{0};
+  std::atomic<int64_t> total_calls{0};
+  std::atomic<int64_t> total_time_us{0};
+  std::atomic<int64_t> max_time_us{0};
+};
+
+// Holder struct for the counter data map
+struct CounterDataMapHolder {
+  c10::Synchronized<
+      std::unordered_map<std::string, std::shared_ptr<CounterData>>>
+      map;
+};
+
+// Leaky singleton to avoid static destruction order issues
+CounterDataMapHolder* getCounterDataMapHolder() {
+  static CounterDataMapHolder* holder = new CounterDataMapHolder();
+  return holder;
+}
+
+// Backend implementation that tracks counter metrics
+class TrackingBackend : public c10::monitor::detail::WaitCounterBackendIf {
+ public:
+  explicit TrackingBackend(std::string key) : key_(std::move(key)) {
+    // Get or create counter data for this key
+    getCounterDataMapHolder()->map.withLock([&](auto& map) {
+      auto it = map.find(key_);
+      if (it == map.end()) {
+        data_ = std::make_shared<CounterData>();
+        map[key_] = data_;
+      } else {
+        data_ = it->second;
+      }
```

Comment:
This entire code can be replaced and more made efficient setting `data_` equal to the return value of try_emplace.second on the `map`. Returns the newly created CounterData if it was created by emplace or the existing data in the map. Have try_emplace handle the implicit construction of shared pointer of Counter. Only one lookup of key_ is needed instead of two and the shared_pointer is constructed in place in the map only if needed.
```suggestion
      data_ = map.try_emplace(key_).second.second
```